### PR TITLE
Feature: NonFungible Refactor

### PR DESCRIPTION
**Pull request by talekhinezh**
## Summary
Refactor NonFungibleData handling, adding references and schema checks.

## Details
* Add schema check to NonFungibleData and KeyValueStores
* Represent NonFungible Data as a single struct and allow references to be stored (disallow ownership)
* Change update non fungible data interface
* Remove NonFungibleData Substate Type and Offsets, use KeyValueStoreEntry for these
* Split Fungible and NonFungible ResourceManager Blueprint and substates

### For Dapp Developers
Creating NonFungible Resources in Scrypto requires user to specify the NonFungible data type:
```
#[derive(ScryptoSbor, NonFungibleData)]
struct NFData {
    pub name: String,
    #[mutable]
    pub flag: bool,
}

{
    ResourceBuilder::new_uuid_non_fungible::<NFData>()
    ...
}
```

Changed update non_fungible data interface:
```
#[derive(Debug, PartialEq, Eq, ScryptoSbor, NonFungibleData)]
pub struct Sandwich {
    pub name: String,
    #[mutable]
    pub available: bool,
}

borrow_resource_manager!(resource_address).update_non_fungible_data(
     non_fungible_local_id,
     "available",
     true,
);
```

### For Internal Integrators
* Split `ResourceManagerSubstate` to `FungibleResourceManagerSubstate` and `NonFungibleResourceManagerSubstate`
* Remove `NonFungibleEntry` Substate/offsets
* Update TypeInfo substate to include KeyValueStore schema

### CODE REVIEW

File: assets/faucet.wasm
Comment (iamyulong) [inline assets/faucet.wasm:1]: Wow, the schema generation code isn't small.

File: radix-engine/src/blueprints/resource/non_fungible_resource_manager.rs [hunk 1]
```diff
@@ -0,0 +1,845 @@
...
+    }
+
+    // Key
+    kv_schema
+        .type_validations
+        .extend(key_schema.type_validations);
+
+    // Optional value
+    kv_schema.type_validations.push(TypeValidation::None);
+    let value_index = LocalTypeIndex::SchemaLocalIndex(kv_schema.type_validations.len() - 1);
```
Comment (iamyulong) [inline radix-engine/src/blueprints/resource/non_fungible_resource_manager.rs:91]: It's slightly harder than I expected to extend an Option. Nice work!

**Review by iamyulong — COMMENTED** [review]
Solid work!

File: radix-engine-tests/tests/non_fungible.rs [hunk 2]
```diff
@@ -1,3 +1,5 @@
+use radix_engine::blueprints::resource::NonFungibleResourceManagerError;
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine-tests/tests/non_fungible.rs:1]: Yup, will do

File: radix-engine/src/blueprints/resource/resource_manager_util.rs [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+
+    api.globalize_with_address(
+        RENodeId::Object(object_id),
+        btreemap!(
+            NodeModuleId::AccessRules => resman_access_rules.id(),
+            NodeModuleId::AccessRules1 => vault_access_rules.id(),
+            NodeModuleId::Metadata => metadata.id(),
+        ),
+        resource_address.into(),
+    )?;
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/blueprints/resource/resource_manager_util.rs:271]: I think `BTreeMap` still okay. We would just have schema checks at the system layer similar to how we have substate schema checks

File: radix-engine/src/system/client_api_impl.rs [hunk 4]
```diff
@@ -92,9 +90,37 @@ where
...
+        } = self.kernel_get_lock_info(lock_handle)?;
+
+        if module_id.eq(&NodeModuleId::SELF) {
+            let type_info = TypeInfoBlueprint::get_type(node_id, self)?;
+            match type_info {
+                TypeInfoSubstate::KeyValueStore(schema) => {
+                    validate_payload_against_schema(&buffer, &schema.schema, schema.value)
+                        .map_err(|_| RuntimeError::KernelError(KernelError::InvalidOverwrite))?;
+
+                    if !schema.can_own {
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/system/client_api_impl.rs:107]: Ooo yeah that's a good idea. Opens up the possibility of recursion checks then :)

File: radix-engine/src/system/node_modules/type_info/package.rs [hunk 5]
```diff
@@ -3,17 +3,21 @@ use crate::kernel::kernel_api::KernelSubstateApi;
...
-    pub package_address: PackageAddress,
-    pub blueprint_name: String,
-    pub global: bool,
+pub enum TypeInfoSubstate {
+    Object {
+        package_address: PackageAddress,
+        blueprint_name: String,
+        global: bool,
+    },
+    KeyValueStore(KeyValueStoreSchema),
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/system/node_modules/type_info/package.rs:15]: Yup exactly my thoughts, except it's not a variant of RENode. It's just something that the system layer understands.

File: radix-engine/src/system/client_api_impl.rs [hunk 6]
```diff
@@ -567,13 +605,48 @@ where
...
+        };
+
+        Ok(schema)
+    }
+
+    fn new_key_value_store(
+        &mut self,
+        schema: KeyValueStoreSchema,
+    ) -> Result<KeyValueStoreId, RuntimeError> {
         let node_id = self.kernel_allocate_node_id(RENodeType::KeyValueStore)?;
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/system/client_api_impl.rs:642]: Yup

**Review by iamyulong — APPROVED** [review]

**Pull request merged**